docs/src/ref/mcmc.md

```diff
@@ -201,6 +201,207 @@ new_trace = permute_move(trace, 2)
+```julia
+@gen function model()
+    if ({:z} ~ bernoulli(0.5))
+        m1 = ({:m1} ~ gamma(1, 1))
+        m2 = ({:m2} ~ gamma(1, 1))
+    else
+        m = ({:m} ~ gamma(1, 1))
+        (m1, m2) = (m, m)
+    end
+    {:y1} ~ normal(m1, 0.1)
+    {:y2} ~ normal(m2, 0.1)
+end
+```
+Because this model has stochastic control flow, it represents two distinct structural hypotheses about how the observed data could have been generated:
+If `:z` is `true` then we enter the first branch, and we hypothesize that the two data points were generated from separate means, sampled at addresses `:m1` and `:m2`.
+If `:z` is `false` then we enter the second branch, and we hypohesize that there is a single mean that explains both data points, sampled at address `:m`.
+
+We want to construct an MCMC kernel that is able to transition between these two distinct structural hypotheses.
+We could construct such a kernel with the simpler 'selection' variant of Metropolis-Hastings, by selecting the address `:z`, e.g.:
+```julia
+select_mh_structure_kernel(trace) = mh(trace, select(:z))[1]
+```
+Sometimes, this kernel would propose to change the value of `:z`.
+We could interleave this kernel with another kernel that does inference over the mean random choices, without changing the structure, e.g.:
+```julia
+@gen function fixed_structure_proposal(trace)
+    if trace[:z]
+        {:m1} ~ normal(trace[:m1], 0.1)
+        {:m2} ~ normal(trace[:m2], 0.1)
+    else
+        {:m} ~ normal(trace[:m], 0.1)
+    end
+end
+
+fixed_structure_kernel(trace) = mh(trace, fixed_structure_proposal, ())[1]
+```
+Combining these together, and applying to particular data and with a specific initial hypotheses:
+```julia
+(y1, y2) = (1.0, 1.3)
+trace, = generate(model, (), choicemap((:y1, y1), (:y2, y2), (:z, false), (:m, 1.2)))
+for iter=1:100
+    trace = select_mh_structure_kernel(trace)
+    trace = fixed_structure_kernel(trace)
+end
+```
+However, this algorithm will not be very efficient, because the internal proposal used by the selection variant of MH is not specialized to the model.
+In particular, when switching from the model with a single mean to the model with two means, the values of the new addresses `:m1` and `:m2` will be proposed from the prior distribution.
+This is wasteful, since if we have inferred an accurate value for `:m`, we expect the values for `:m1` and `:m2` to be near this value.
+The same is true when proposing a structure change in the opposite direction.
+That means it will take many more steps to get an accurate estimate of the posterior probability distribution on the two structures.
+
+We would like to use inferred values for `:m1` and `:m2` to inform our proposal for the value of `:m`.
+For example, we could take the harmonic mean:
+```julia
+m = sqrt(m1 * m2)
+```
+However, there are many combinations of `m1` and `m2` that have the same harmonic mean.
```

Comment:
This is the geometric mean.

Reply:
Fixed

Reply:
> 3. What is the difference under the hood between reading and copying?  Which of them (if any) corresponds to saying "treat these as constants with their current value when computing the Jacobian"?

A value can appear in the output pair of choice maps in one of three ways:
- explicit write
- explicit copy (copying values from input choice maps to output choice maps, potentially at different addresses)
- implicit copy (this can only occur for model address, and is done implicitly by 'update'; the addresses that are implicitly copied are those that are (i) not deleted from the model choice map, and (ii) not the target of an explicit write, and (iii) not the target of an explicit copy).

So, "treat these as constants with their current value, at their current address" is handled by the "implicit copy". But explicit copies also afford simplification of the Jacobian that needs to be computed, because like implicit copies, the explicit copies also produce columns in the Jacobian that contain all 0s except for a single 1 (which allows them to be ignored for the purposes of computing the determinant).

Under the hood, we first run the involution, and record each read, write, and explicit copy (address). Then, we determine the subset of these commands for which we actually need to compute their corresponding row or column in the Jacobian, and produce a version of the involution (f_array) that takes as input an array containing the values of just those special input addresses, and returns as output an array containing the values of just those special output addresses (the fact that the number of  such outputs and inputs is equal is not obvious, and a proof of this is given in the LaTex notes I sent).

> 4. I have thoughts about extending this framework to allow for functions `h` that are [covering maps](https://en.wikipedia.org/wiki/Covering_space) such that `h ∘ h` is not necessarily the identity but is a [deck transformation](https://en.wikipedia.org/wiki/Covering_space#Deck_(covering)_transformation_group,_regular_covers).  This could help with situations where we can't globally define a canonical branch of the inverse, for example mapping an angle `theta` to a point `(cos(theta), sin(theta))` on the unit circle where we have to branch-cut at some point and wrap around by 2pi.  Definitely future work though.

Interesting. On a related note, I haven't looked into applying this DSL this for quaternion-valued random choices yet. Let me know if you want to discuss that some time. Maybe your generalization is useful there.